### PR TITLE
RavenDB-19186: Dense BitVector and usage of framework primitives for bits handling.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-19186

### Additional description
While technically part of RavenDB-19186 work, this PR is a performance and support classes implementation. 

### Type of change
- Optimization
- New feature

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### CODE REVIEW

File: src/Sparrow/Binary/BitVector.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+#if NET7_0_OR_GREATER
+            if (Avx2.IsSupported)
+                return IndexOfFirstSetBitAvx2(ref storage, _storage.Length);
+            if (Sse2.IsSupported)
+                return IndexOfFirstSetBitSse2(ref storage, _storage.Length);
+#endif
+            return IndexOfFirstSetBitScalar(ref storage, _storage.Length);
+        }
+
+        public string ToString()
```
Comment (arekpalinski) [inline src/Sparrow/Binary/BitVector.cs]: Please check the warning below

**Comment by arekpalinski** [comment]
Do we have any benchmark results? Can we put them here for reference?

**Comment by redknightlois** [comment]
Pointer and Span based bitvector are equal now, the span version was much slower and bigger code because of bounds checks.

```

|     Method |     Mean |    Error |   StdDev | Ratio |
|----------- |---------:|---------:|---------:|------:|
|    Current | 12.00 us | 0.005 us | 0.004 us |  1.00 |
| CurrentPtr | 12.13 us | 0.016 us | 0.013 us |  1.01 |
```

**Review by arekpalinski — APPROVED** [review]

**Comment by maciejaszyk** [comment]
LGTM

I wrote a test for checking accelerated and non-accelerated (requires modifying the BitVector). You can put without that flag:
``` csharp
  [Theory]
        [InlineData(true, 3)]
        [InlineData(false, 3)]
        [InlineData(true, 256)]
        [InlineData(false, 256)]
        [InlineData(true, 1024)]
        [InlineData(false, 1024)]
        [InlineData(false, 1025)]
        [InlineData(true, 1025)]

        public void BitVectorTest(bool isAccelerated, int size)
        {
            if (isAccelerated && (Avx2.IsSupported || Sse2.IsSupported) == false)
                return;

            var buffer = new byte[size];
            buffer.AsSpan().Clear();

            var bitVector = new BitVector(buffer, isAccelerated == false);
            
            Assert.Equal(-1, bitVector.IndexOfFirstSetBit());

            var random = new Random(12532423);
            var bufferCopy = new bool[size];
            
            
            for (int i = 0; i < size; ++i)
            {
                var id = random.Next(0, size);
                Assert.Equal(bufferCopy[id], bitVector[id]);
                bufferCopy[id] = bufferCopy[id] == false;

                bitVector[id] = bitVector[id] == false;
                AssertBitVectorAndBufferCopy(in bitVector);
            }

            void AssertBitVectorAndBufferCopy(in BitVector bitVector)
            {
                for (int i = 0; i < size; ++i)
                    Assert.Equal(bufferCopy[i],bitVector[i]);
            }
        }
        ```

**Review by maciejaszyk — APPROVED** [review]

**Comment by redknightlois** [comment]
> LGTM
> 
> I wrote a test for checking accelerated and non-accelerated (requires modifying the BitVector). You can put without that flag:
> 
> ```cs
>   [Theory]
>         [InlineData(true, 3)]
>         [InlineData(false, 3)]
>         [InlineData(true, 256)]
>         [InlineData(false, 256)]
>         [InlineData(true, 1024)]
>         [InlineData(false, 1024)]
>         [InlineData(false, 1025)]
>         [InlineData(true, 1025)]
> 
>         public void BitVectorTest(bool isAccelerated, int size)
>         {
>             if (isAccelerated && (Avx2.IsSupported || Sse2.IsSupported) == false)
>                 return;
> 
>             var buffer = new byte[size];
>             buffer.AsSpan().Clear();
> 
>             var bitVector = new BitVector(buffer, isAccelerated == false);
>             
>             Assert.Equal(-1, bitVector.IndexOfFirstSetBit());
> 
>             var random = new Random(12532423);
>             var bufferCopy = new bool[size];
>             
>             
>             for (int i = 0; i < size; ++i)
>             {
>                 var id = random.Next(0, size);
>                 Assert.Equal(bufferCopy[id], bitVector[id]);
>                 bufferCopy[id] = bufferCopy[id] == false;
> 
>                 bitVector[id] = bitVector[id] == false;
>                 AssertBitVectorAndBufferCopy(in bitVector);
>             }
> 
>             void AssertBitVectorAndBufferCopy(in BitVector bitVector)
>             {
>                 for (int i = 0; i < size; ++i)
>                     Assert.Equal(bufferCopy[i],bitVector[i]);
>             }
>         }
>         ```
> ```

No problem, just commit what's necessary and I will incorporate it into the PR.

**Pull request merged**